hashing/sha256.cpp

```diff
@@ -0,0 +1,258 @@
+/**
+ * @file
+ * @author [Md. Anisul Haque](https://github.com/mdanisulh)
+ * @brief Simple C++ implementation of the [SHA-256 Hashing Algorithm]
+ * (https://en.wikipedia.org/wiki/SHA-2)
+ *
+ * @details
+ * [SHA-2](https://en.wikipedia.org/wiki/SHA-2) is a set of cryptographic hash
+ * functions that was designed by the
+ * [NSA](https://en.wikipedia.org/wiki/National_Security_Agency) and first
+ * published in 2001. SHA-256 is a part of the SHA-2 family. SHA-256 is widely
+ * used for authenticating software packages and secure password hashing.
+ */
+
+#include <array>     /// For std::array
+#include <cassert>   /// For assert
+#include <cstdint>   /// For uint8_t, uint32_t and uint64_t data types
+#include <iomanip>   /// For std::setfill and std::setw
+#include <iostream>  /// For IO operations
+#include <sstream>   /// For std::stringstream
+#include <vector>    /// For std::vector
+
+/**
+ * @namespace hashing
+ * @brief Hashing algorithms
+ */
+namespace hashing {
+/**
+ * @namespace SHA-256
+ * @brief Functions for the [SHA-256](https://en.wikipedia.org/wiki/SHA-2)
+ * algorithm implementation
+ */
+namespace sha256 {
+/**
+ * @brief Returns the character at pos after the input is padded
+ * @param input Input string
+ * @param pos Position of character to be returned
+ * @param padded_input_size Size of the padded input
+ * @return char Character at the index pos in the padded string
+ */
+char get_char(const std::string &input, std::size_t pos,
+              size_t padded_input_size) {
+    size_t input_size = input.length();
+    char ch = 0;
+    if (pos < input_size) {
+        ch = input[pos];
+    } else if (pos == input_size) {
+        ch = '\x80';
+    } else if (pos < padded_input_size - 8) {
+        ch = '\x00';
+    } else {
+        ch =
+            static_cast<char>((input_size * 8 >> (56 - (pos % 56) * 8)) & 0xFF);
+    }
+    return ch;
+}
```

Comment:
I am not sure if ```assert```s is actually required because ```pos``` and ```byte_num``` will always be in range otherwise one of the previous if conditions will become true.

Reply:
What if someone would call this function like `get_char(some_str, 10000, 10)`? In this cane one can argue that this function is so _specific_ that it will not happen. But what about `extract_byte(some_64_bit_uint, 10)`? What is the 10th byte of the 64 bit variable? It is _zero_ or does it not exist? Maybe by throwing an exception, one can prevent a hard to find bug.